File: Misc/Kosaraju.js
// node --max-stack-size 32000 Kosaraju.js input/SCC.txt
var fs = require('fs');

var inputFile = process.argv[2],
    text = fs.readFileSync(inputFile, 'utf-8'),
		input = text.split('\n'),
    len = input.length, i;

// note: the input reflects that this is an directed graph
var graph = {},
		largest = -Infinity, i;
for (i = 0; i < len; i++) {
  var edge = input[i].trim().split(' ');
  if (!(edge[0] in graph)) {
  	graph[edge[0]] = [];
  }

  // these if statements are here for the case when the largest numbered vertex(s) in the original graph don't have outgoing edges
  if (parseInt(edge[0]) > largest) {
  	largest = parseInt(edge[0]);
  }
  if (parseInt(edge[1]) > largest) {
  	largest = parseInt(edge[1]);
  }

  graph[edge[0]].push(parseInt(edge[1], 10));
}
//console.log(graph)
console.log('Graph built')

// reverse graph for 1st DFS pass
var reversedGraph = {};
for (var vertex in graph) {
	var neighbors = graph[vertex];
	for (i = 0; i < neighbors.length; ++i) {
		if (!(neighbors[i] in reversedGraph)) {
			reversedGraph[neighbors[i]] = [];
		}
		reversedGraph[neighbors[i]].push(vertex);
	}
}
console.log('Reversed graph');

function dfs(graph, vertex) {
	if (visited[vertex]) {
		return;
	}

	visited[vertex] = true;
	//console.log('Visiting:', vertex);
	leaders[vertex] = s;

	var neighbors = graph[vertex];
	if (neighbors !== undefined) {
		for (var i = 0; i < neighbors.length; i++) {
			dfs(graph, neighbors[i]);
		}
	}
	
	rank += 1;
	finish[vertex] = rank;
}

var visited = {},
		leaders = {},
		finish = {},
		rank = 0,
		vertices = largest,	//Object.keys(graph).length,
		s;	// used to find the leader of strongly connected components

// 1st DFS-loop
for (i = vertices; i > 0; --i) {
	s = i;	// set the leader
	dfs(reversedGraph, i);
}
console.log('Completed 1st DFS loop')

// replace original graph vertices with finish #'s
var newGraph = {};
for (i = 1; i <= vertices; ++i) {
	if (graph[i]) {
		newGraph[finish[i]] = graph[i];

		// update the neighbors as well
		var neighbors = graph[i];
		for (var j = 0; j < neighbors.length; j++) {
			if (finish[neighbors[j]]) {
				neighbors[j] = finish[neighbors[j]];
			}
		}
	}
}
console.log('Created new graph');

// 2nd DFS-loop
rank = 0;
visited = {};
leaders = {};
for (i = vertices; i > 0; --i) {
	s = i;	// set the leader
	dfs(newGraph, i);
}
console.log('Completed 2nd DFS loop')

var counts = {};
for (vertex in leaders) {
	if (!counts[leaders[vertex]]) {
		counts[leaders[vertex]] = 0;
	}
	counts[leaders[vertex]] += 1;
}
//console.log(counts);

var sortable = [];
for (vertex in counts) {
  sortable.push([vertex, counts[vertex]]);
}
sortable.sort(function(a, b) { return a[1] - b[1]; });
console.log(sortable);

File: Misc/Kosaraju.py
file = open('test.txt', 'r')

def dfs(graph, vertex, visited, finish, leaders, rank, s):
	if (vertex in visited):
		return

	visited[vertex] = True
	print 'Visiting:', vertex
	leaders[vertex] = s

	if vertex in graph:
		neighbors = graph[vertex];
		for i in xrange(len(neighbors)):
			dfs(graph, neighbors[i], visited, finish, leaders, rank, s);
		
	rank['rank'] += 1
	finish[vertex] = rank['rank']

def iterative_dfs(graph, vertex, visited, finish, leaders, rank, s):
	if (vertex in visited):
		return

	open = [vertex]
	while (len(open) > 0):
		print 'Visiting:', vertex
		vertex = open[0]
		open.pop(0)
		leaders[vertex] = s
		visited[vertex] = True

		if vertex in graph:
			neighbors = graph[vertex]
			for i in neighbors:
				if i not in visited:
					open.append(i)

graph = {}
for line in file:
	i = line.strip().split(' ')
	tail, head = int(i[0]), int(i[1])
	if tail not in graph:
		graph[tail] = []
	graph[tail].append(head)
print 'Successfully read file and created graph!'

reversedGraph = {};
for vertex in graph:
	neighbors = graph[vertex];
	for i in xrange(len(neighbors)):
		if (neighbors[i] not in reversedGraph):
			reversedGraph[neighbors[i]] = [];	
		reversedGraph[neighbors[i]].append(int(vertex));
print 'Successfully reversed graph!'

visited = {}
leaders = {}
finish = {}
rank = {}
rank['rank'] = 0
s = None	# used to find the leader of strongly connected components

# 1st DFS-loop
for i in xrange(len(graph), -1, -1):
	vertex = i + 1
	s = vertex	# set the leader
	iterative_dfs(reversedGraph, vertex, visited, finish, leaders, rank, s)

print 'Completed first DFS-loop'
print leaders
print finish
'''
# replace original graph vertices with finish #'s
newGraph = {};
print graph
for i in xrange(len(graph), -1, -1):
	vertex = i + 1
	if vertex in graph:
		newGraph[finish[vertex]] = graph[vertex];
print 'Used finish array to change vertex names'

# reset global vars
visited = {}
leaders = {}
finish = {}
rank = {}
rank['rank'] = 0
s = None

# 2nd DFS-loop
for i in xrange(len(graph), -1, -1):
	vertex = i + 1
	s = vertex	# set the leader
	dfs(newGraph, vertex, visited, finish, leaders, rank, s)
print 'Completed second DFS-loop'
print leaders
'''